os/kernel/sched/sched_addDeadtask.c

```diff
@@ -0,0 +1,255 @@
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sched_deadtasklistinit
+ *
+ * Description:
+ *   This function initializes dead task info list
+ *
+ * Inputs:
+ *   None
+ *
+ ****************************************************************************/
+
+void sched_deadtasklistinit()
+{
+    sq_init(&dead_pidList);
+    sq_init(&dead_nameList);
+}
+
+/****************************************************************************
+ * Name: sched_addDeadtask
+ *
+ * Description:
+ *   This function adds a task's pid and name info to the list.
+ *   This should be called before deleting a task.
+ *
+ * Inputs:
+ *   pid: pid of the deleting task
+ *   name: name of the task
+ *
+ ****************************************************************************/
+
+void sched_addDeadtask(int pid, char *name)
+{
+    dead_pid_t *pid_head = (dead_pid_t *)dead_pidList.head;
+    dead_pid_t *pid_node = NULL;
+
+    irqstate_t flags = enter_critical_section();
+
+    /* Insert pid node in pid list */
+    if (!pid_head) {
+        pid_node = allocate_pid_node(pid);
+        sq_addfirst((sq_entry_t *)pid_node, &dead_pidList);
+    } else {
+        dead_pid_t *pid_prev;
+        while (pid_head) {
+            if (pid_head->pid == pid) {
+                dbg("ERROR: pid is already present in the list of dead pid\n");
+                return;
+            }
+            pid_prev = pid_head;
+            pid_head = pid_head->next;
+        }
+        pid_node = allocate_pid_node(pid);
+        sq_addafter((sq_entry_t *)pid_prev, (sq_entry_t *)pid_node, &dead_pidList);
+    }
+
+    /* Insert name node if the name not present in name list already */
+    dead_name_t *name_head = (dead_name_t *)dead_nameList.head;
+    dead_name_t *name_node = NULL;
+
+    if (!name_head) {
+        name_node = allocate_name_node(name);
+        sq_addfirst((sq_entry_t *)name_node, &dead_nameList);
+    } else {
+        dead_name_t *name_prev;
+        while (name_head) {
+            if (strcmp(name_head->name, name) == 0) {
+                /* point the pid node to this name_node */
+                pid_node->name_ptr = name_head;
+                name_head->pid_cnt++;
+                return;
+            }
+            name_prev = name_head;
+            name_head = name_head->next;
+        }
+        name_node = allocate_name_node(name);
+        sq_addafter((sq_entry_t *)name_prev, (sq_entry_t *)name_node, &dead_nameList);
+    }
+
+    leave_critical_section(flags);
+
+    /* point to newly created name node */
+    pid_node->name_ptr = name_node;
+    name_node->pid_cnt++;
+    return;
+}
+
+/****************************************************************************
+ * Name: sched_removedeadtask
+ *
+ * Description:
+ *   This function removes a task's pid and name info from dead list
+ *   if present. This should be called after creating a task.
```

Comment:
Wont this create confusion if we are only printing the most recent name of that pid,

at time1, pid1 is task1, it reserved some memory and terminated, but didnt free it.

at time2, pid1 is reassigned to task2, now according to the above logic, when dead threads list is printed, the memory allocated by task1 will be wrongly shown as allocated by task2.

Reply:
I think that is ok. The max pid count in our system is 65535 (16 bit pid_t). If this is exhausted, then our counter will roll over and start checking from zero. It will allocate a pid of a previously dead task for the new task. If we want to completely avoid the situation that you mentioned, then we need to keep storing all previous task names also, I think it might be an overkill. 

@sunghan-chang Please let us know your opinion.